### PR TITLE
Fix for not wrapping a function for a different Wrappee.

### LINKED ISSUE #146: Gotcha breaks OpenCL
I have a reproducer that breaks valid OpenCL applications when wrapping OpenCL functions with GOTCHA

Compile either with or without `-DUSE_GOTCHA`:

```c
/* -*- c -*- */

#define _GNU_SOURCE
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <stdbool.h>
#include <math.h>

#ifdef __APPLE__
#  include <OpenCL/opencl.h>
#else
#  include <CL/cl.h>
#endif

#ifdef USE_GOTCHA

#include <gotcha/gotcha.h>

// We need a place to store the pointer to the function we've wrapped
static gotcha_wrappee_handle_t handle_clGetPlatformIDs;
static gotcha_wrappee_handle_t handle_clGetExtensionFunctionAddress;
static gotcha_wrappee_handle_t handle_clGetPlatformInfo;

static cl_int
doClGetPlatformIDs(cl_uint         numEntries,
                   cl_platform_id* platforms,
                   cl_uint*        numPlatforms)
{
    fprintf(stderr, "E clGetPlatformIDs(%u)\n", numEntries);

    typeof(&doClGetPlatformIDs) origClGetPlatformIDs = gotcha_get_wrappee(handle_clGetPlatformIDs);
    cl_int ret = origClGetPlatformIDs(numEntries, platforms, numPlatforms);

    fprintf(stderr, "E clGetPlatformIDs(%u) = %d\n", numEntries, ret);

    return ret;
}

static void*
doClGetExtensionFunctionAddress( const char* funcName )
{
    fprintf(stderr, "E clGetExtensionFunctionAddress(%s)\n", funcName);

    typeof(&doClGetExtensionFunctionAddress) origClGetExtensionFunctionAddress = gotcha_get_wrappee(handle_clGetExtensionFunctionAddress);
    void* ret = origClGetExtensionFunctionAddress( funcName );

    fprintf(stderr, "L clGetExtensionFunctionAddress(%s) = %p\n", funcName, ret);

    return ret;
}

static cl_int
doClGetPlatformInfo(cl_platform_id   platform,
                    cl_platform_info paramName,
                    size_t           paramValueSize,
                    void*            paramValue,
                    size_t*          paramValueSizeRet)
{
    fprintf(stderr, "E clGetPlatformInfo()\n");

    typeof(&doClGetPlatformInfo) origClGetPlatformInfo = gotcha_get_wrappee(handle_clGetPlatformInfo);
    cl_int ret = origClGetPlatformInfo(platform, paramName, paramValueSize, paramValue, paramValueSizeRet);

    fprintf(stderr, "L clGetPlatformInfo()\n");

    return ret;
}

struct gotcha_binding_t bindings[] = {
    { "clGetPlatformIDs", doClGetPlatformIDs, &handle_clGetPlatformIDs},
    { "clGetExtensionFunctionAddress", doClGetExtensionFunctionAddress, &handle_clGetExtensionFunctionAddress},
    { "clGetPlatformInfo", doClGetPlatformInfo, &handle_clGetPlatformInfo}
};

#endif

int
main(int ac, char *av[])
{
#ifdef USE_GOTCHA
    gotcha_wrap(bindings, 2, "ocl_test");
#endif

    cl_uint platform_count = 0;

    clGetPlatformIDs(0, NULL, &platform_count);

    if (platform_count == 0) {
        printf("No OpenCL platforms found!\n");
        return 127;
    } else {
        cl_platform_id* platforms = calloc(platform_count, sizeof(*platforms));
        clGetPlatformIDs(platform_count, platforms, NULL);

        printf("Platforms:\n");

        for (size_t i = 0; i < platform_count; i++) {
            char platform_name[256];
            clGetPlatformInfo(platforms[i], CL_PLATFORM_NAME, sizeof(platform_name), platform_name, NULL);
            printf(" %zu: %s\n", i, platform_name);
        }

        free(platforms);
    }

    return 0;
}
```

When running the non-wrapped version and setting a breakpoint at `clGetExtensionFunctionAddress`, the call chain looks like this:

```pascal
#0  0x00007ffff7c64400 in clGetExtensionFunctionAddress () from /opt/rocm-6.0.2/lib/libamdocl64.so
#1  0x00007ffff7f8e532 in ?? () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#2  0x00007ffff7f9058c in ?? () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#3  0x00007ffff7d874df in __pthread_once_slow (once_control=0x7ffff7f940f8, init_routine=0x7ffff7f90420) at pthread_once.c:116
#4  0x00007ffff7f8eb95 in clGetPlatformIDs () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#5  0x000000000020190f in main ()
```

`clGetExtensionFunctionAddress` is called from inside `libOpenCL.so.1` but will be dispatched to the AMD OpenCL ICD `libamdocl64.so`.

When doing the same with the wrapped binary:

```pascal
#0  0x00007ffff7f8fc60 in clGetExtensionFunctionAddress () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#1  0x0000000000201c24 in doClGetExtensionFunctionAddress ()
#2  0x00007ffff7f8e532 in ?? () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#3  0x00007ffff7f9058c in ?? () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#4  0x00007ffff7d764df in __pthread_once_slow (once_control=0x7ffff7f940f8, init_routine=0x7ffff7f90420) at pthread_once.c:116
#5  0x00007ffff7f8eb95 in clGetPlatformIDs () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#6  0x0000000000201bb1 in doClGetPlatformIDs ()
#7  0x0000000000201d3d in main ()
```

the wrapper calls `clGetExtensionFunctionAddress` from `libOpenCL.so.1` and not from `libamdocl64.so`. But because the OpenCL library is still in its initialization, `clGetExtensionFunctionAddress` tries to init itself again but that deadlocks because of the recursive call to `pthread_once`. Here is the backtrace when continuing:

```pascal
#0  futex_wait (private=0, expected=1, futex_word=0x7ffff7f940f8) at ../sysdeps/nptl/futex-internal.h:141
#1  futex_wait_simple (private=0, expected=1, futex_word=0x7ffff7f940f8) at ../sysdeps/nptl/futex-internal.h:172
#2  __pthread_once_slow (once_control=0x7ffff7f940f8, init_routine=0x7ffff7f90420) at pthread_once.c:105
#3  0x00007ffff7f8fc7f in clGetExtensionFunctionAddress () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#4  0x0000000000201c24 in doClGetExtensionFunctionAddress ()
#5  0x00007ffff7f8e532 in ?? () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#6  0x00007ffff7f9058c in ?? () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#7  0x00007ffff7d764df in __pthread_once_slow (once_control=0x7ffff7f940f8, init_routine=0x7ffff7f90420) at pthread_once.c:116
#8  0x00007ffff7f8eb95 in clGetPlatformIDs () from /opt/rocm-6.0.2/lib/libOpenCL.so.1
#9  0x0000000000201bb1 in doClGetPlatformIDs ()
#10 0x0000000000201d3d in main ()
```

You will notice that the arguments to `__pthread_once_slow` are the same in level `#2` and `#7`.

My current work around is to filter `libOpenCL.so`.

### PR DESCRIPTION
**Pull request by hariharan-devarajan**
In cases, like in OpenCL, the main library (OpenCL) acts as a wrapper for other implementations (libamdocl64.so). In this case both have same function names implemented. If we wrap OpenCL functions with gotcha which dynamically loads libamdocl64.so with same function name, we should not return a wrapper meant to intercept OpenCL. So we should check if the wrapper given to the caller of dlsym is for the same wrappee as requested by the caller.

Fixes #146. @bertwesarg So essentially, what was happening is 

When u call `gotcha_wrap`, u wrapped the `clGetExtensionFunctionAddress` function within libOpenCL.so.

Now the OpenCL, `dlopen` libamdocl64.so and did a dlsym for `clGetExtensionFunctionAddress`. GOTCHA was returning the wrapper for libOpenCL.so. 

The fix essentially adds a check to make sure we return the wrapper for the function we wrapped. Let me know if this works.

### CODE REVIEW

File: test/dispatcher/libdispatcher.c [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+static int (*impl_foo)(void);
+static int (*impl_bar)(void);
+
+static pthread_once_t init_once = PTHREAD_ONCE_INIT;
+void dispatch_init(void) {
+  fprintf(stderr, "Ed dispatch_init()\n");
+
+  impl_lib = dlopen("libimpl.so", RTLD_NOW);
+  impl_foo = dlsym(impl_lib, "foo");
+  impl_bar = dlsym(impl_lib, "bar");
```
Comment (bertwesarg) [inline test/dispatcher/libdispatcher.c:40]: please add some error checking here. when running without `LD_LIBRARY_PATH` set, the test hangs again bacause the subsequent `dlsym` calls return the symbols from the dispatcher library again

```suggestion
  impl_lib = dlopen("libimpl.so", RTLD_NOW);
  assert(impl_lib);
  impl_foo = dlsym(impl_lib, "foo");
  assert(impl_foo);
  impl_bar = dlsym(impl_lib, "bar");
  assert(impl_bar);
```

File: test/dispatcher/libdispatcher.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#include <dlfcn.h>
+#include <math.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
```
Comment (bertwesarg) [inline test/dispatcher/libdispatcher.c:26]: thats just my default set of `#include` _s_, we should be able to reduce this

```suggestion
#define _GNU_SOURCE
#include <dlfcn.h>
#include <pthread.h>
#include <stdio.h>
```

File: test/dispatcher/libimpl.c [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+#define _GNU_SOURCE
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
```
Comment (bertwesarg) [inline test/dispatcher/libimpl.c:23]:
```suggestion
#include <stdio.h>
```

File: test/dispatcher/main.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+*/
+
+#define _GNU_SOURCE
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
```
Comment (bertwesarg) [inline test/dispatcher/main.c:23]:
```suggestion
#define _GNU_SOURCE
#include <stdio.h>
```

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -12,4 +12,5 @@ add_subdirectory(filter)
 add_subdirectory(wrap_main)
 #add_subdirectory(multi_agent_dlopen)
 add_subdirectory(symver)
-add_subdirectory(function_ptr)
\ No newline at end of file
+add_subdirectory(function_ptr)
+add_subdirectory(dispatcher)
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply test/CMakeLists.txt]: FIXED

File: .gitignore [hunk 6]
```diff
@@ -4,3 +4,7 @@
 build
 tags
 **/tags
+*.core
+.cache
+.idea
+.vscode
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply .gitignore]: FIXED

File: test/dispatcher/libdispatcher.c (hunk 1 above)
Comment (hariharan-devarajan) [reply test/dispatcher/libdispatcher.c]: FIXED

File: test/dispatcher/libdispatcher.c (hunk 2 above)
Comment (hariharan-devarajan) [reply test/dispatcher/libdispatcher.c]: FIXED

File: test/dispatcher/libimpl.c (hunk 3 above)
Comment (hariharan-devarajan) [reply test/dispatcher/libimpl.c]: FIXED

File: test/dispatcher/main.c (hunk 4 above)
Comment (hariharan-devarajan) [reply test/dispatcher/main.c]: FIXED

**Comment by hariharan-devarajan** [comment]
@bertwesarg did u get a change to look at this PR? Does this cover all the cases you need for #146

**Comment by bertwesarg** [comment]
I will give this PR a last test in my use case and will approve then

**Comment by bertwesarg** [comment]
> I will give this PR a last test in my use case and will approve then

LGTM

**Pull request merged**